warehouse/integrations/github/utils.py

```diff
@@ -0,0 +1,351 @@
+
+        if self.public_keys_cached_at + PUBLIC_KEYS_CACHE_TIME < time.time():
+            raise CacheMiss
+
+        return self.public_keys_cache
+
+    def _headers_auth(self):
+        if not self._api_token:
+            return {}
+        return {"Authorization": f"token {self._api_token}"}
+
+    def _retrieve_public_key_payload(self):
+
+        token_scanning_pubkey_api_url = (
+            "https://api.github.com/meta/public_keys/token_scanning"
+        )
+
+        try:
+            response = self._session.get(
+                token_scanning_pubkey_api_url, headers=self._headers_auth()
+            )
+            response.raise_for_status()
+            return response.json()
+        except requests.HTTPError as exc:
+            raise GitHubPublicKeyMetaAPIError(
+                f"Invalid response code {response.status_code}: {response.text[:100]}",
+                f"public_key_api.status.{response.status_code}",
+            ) from exc
+        except json.JSONDecodeError as exc:
+            raise GitHubPublicKeyMetaAPIError(
+                f"Non-JSON response received: {response.text[:100]}",
+                "public_key_api.invalid_json",
+            ) from exc
+        except requests.RequestException as exc:
+            raise GitHubPublicKeyMetaAPIError(
+                "Could not connect to GitHub", "public_key_api.network_error"
+            ) from exc
+
+    def _extract_public_keys(self, pubkey_api_data):
+        if not isinstance(pubkey_api_data, dict):
+            raise GitHubPublicKeyMetaAPIError(
+                f"Payload is not a dict but: {str(pubkey_api_data)[:100]}",
+                "public_key_api.format_error",
+            )
+        try:
+            public_keys = pubkey_api_data["public_keys"]
+        except KeyError:
+            raise GitHubPublicKeyMetaAPIError(
+                "Payload misses 'public_keys' attribute", "public_key_api.format_error"
+            )
+
+        if not isinstance(public_keys, list):
+            raise GitHubPublicKeyMetaAPIError(
+                "Payload 'public_keys' attribute is not a list",
+                "public_key_api.format_error",
+            )
+
+        expected_attributes = {"key", "key_identifier"}
+        for public_key in public_keys:
+
+            if not isinstance(public_key, dict):
+                raise GitHubPublicKeyMetaAPIError(
+                    f"Key is not a dict but: {public_key}",
+                    "public_key_api.format_error",
+                )
+
+            attributes = set(public_key)
+            if not expected_attributes <= attributes:
+                raise GitHubPublicKeyMetaAPIError(
+                    "Missing attribute in key: "
+                    f"{sorted(expected_attributes - attributes)}",
+                    "public_key_api.format_error",
+                )
+
+            yield {"key": public_key["key"], "key_id": public_key["key_identifier"]}
+
+        self.public_keys_cache = public_keys
```

Comment:
Oh, this line is only executed when we finish the generator, but the generator is evaluated lazily, so we likely never reach this line and consequently alway cache miss :o

Reply:
fixed in ca804ee